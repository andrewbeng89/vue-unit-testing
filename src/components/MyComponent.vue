<!-- Vue component using the mapped getter -->
<template>
  <div>
    <span> The first item is {{ getItem(0) }} </span>
    <button id="btn" @click="onClick">
      Click Me!
    </button>
    <button @click="addItem({ foo: 'bar' })">
      Add Item
    </button>
    <pre>{{ secret }}</pre>
  </div>
</template>

<script>
import Vue from "vue";
import { mapGetters, mapMutations } from "vuex";

export default Vue.extend({
  props: {
    secret: {
      type: String,
      required: false,
      default: "No secret..."
    }
  },
  computed: {
    ...mapGetters(["getItem"])
  },
  methods: {
    ...mapMutations(["addItem"]),
    onClick() {
      this.$emit("my-event");
    }
  }
});
</script>
