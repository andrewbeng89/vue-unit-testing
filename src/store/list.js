import axios from "axios";

// A simple vuex store: list.js
const state = { list: [], status: "" };

const mutations = {
  addItem: (state, item) => {
    state.list = [...state.list, item];
  },
  setItems: (state, items) => {
    state.list = items;
  },
  addItems: (state, items) => {
    state.list = [...state.list, ...items];
  },
  setStatus: (state, status) => {
    state.status = status;
  }
};

const actions = {
  fetchItems: async ({ commit }, page) => {
    try {
      const url =
        process.env.NODE_ENV === "development"
          ? "http://127.0.0.1:3000"
          : "https://nodejs-koa.andrewbeng89.now.sh";
      commit("setStatus", "loading");
      const { data } = await axios.get(`${url}/items?page=${page}`);
      const action = page ? "addItems" : "setItems";
      commit(action, data);
      commit("setStatus", "success");
    } catch (error) {
      commit("setStatus", error.message);
    }
  }
};

const getters = {
  getItem: state => index => state.list[index]
};

export default {
  state,
  mutations,
  getters,
  actions
};
