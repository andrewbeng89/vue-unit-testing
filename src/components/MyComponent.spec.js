import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import list from "@/store/list";
import MyComponent from "@/components/MyComponent.vue";

const localVue = createLocalVue();
localVue.use(Vuex);
const store = new Vuex.Store({
  ...list,
  state: {
    ...list.state,
    list: [{ foo: "bar" }]
  }
});

describe("MyComponent.vue", () => {
  it("renders props.secret when passed", () => {
    const secret = "my secret message";
    const wrapper = shallowMount(MyComponent, {
      propsData: { secret },
      store,
      localVue
    });
    expect(wrapper.find("pre").text()).toMatch(secret);
  });

  it("emits 'my-event' if btn is clicked", () => {
    const wrapper = shallowMount(MyComponent, {
      store,
      localVue
    });
    const btn = wrapper.find("#btn");
    btn.trigger("click");
    expect(wrapper.emitted()["my-event"].length).toBe(1);
  });

  it("prints the first item, two buttons, and 'secret'", () => {
    const wrapper = shallowMount(MyComponent, {
      store,
      localVue
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
