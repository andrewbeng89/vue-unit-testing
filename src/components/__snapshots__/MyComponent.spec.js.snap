// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent.vue passes snapshot testing 1`] = `
<div>
  <span>
     The first item is {
  "foo": "bar"
} 
  </span>
   
  <button
    id="btn"
  >
    
    Click Me!
  
  </button>
   
  <button>
    
    Add Item
  
  </button>
   
  <pre>
    No secret...
  </pre>
</div>
`;
